picocrt/machine/arm/crt0.c

```diff
@@ -132,6 +197,55 @@ _start(void)
 	__asm__("vmsr fpexc, %0" : : "r" (0x40000000));
 #endif
 
+#ifdef _PICOCRT_ENABLE_MMU
+
+#if __ARM_ARCH >= 7 && __ARM_ARCH_PROFILE != 'R'
+
+#define SCTLR_MMU (1 << 0)
+#define SCTLR_DATA_L2 (1 << 2)
+#define SCTLR_BRANCH_PRED (1 << 11)
+#define SCTLR_ICACHE (1 << 12)
+
+        uint32_t        mmfr0;
+        __asm__("mrc p15, 0, %0, c0, c1, 4" : "=r" (mmfr0));
+
+        /* Check to see if the processor supports VMSAv7 or better */
+        if ((mmfr0 & 0xf) >= 3)
+        {
+                /* We have to set up an identity map and enable the MMU for caches.
+                 * Additionally, all page table entries are set to Domain 0, so set up DACR
+                 * so that Domain zero has permission checks enabled rather than "deny all".
+                 */
+
+                /* Set DACR Domain 0 permissions checked */
+                __asm__("mcr p15, 0, %0, c3, c0, 0\n" :: "r" (1));
+
+                /*
+                 * Write TTBR
+                 *
+                 * No DSB since tables are statically initialized and dcache is off.
+                 * We or __identity_page_table with 0x3 to set the cacheable flag bits.
+                 */
+                __asm__("mcr p15, 0, %0, c2, c0, 0\n"
+                        :: "r" ((uintptr_t)__identity_page_table | 0x3));
+
+                /* Note: we assume Data+L2 cache has been invalidated by reset. */
+                __asm__("mcr p15, 0, %0, c7, c5, 0\n" :: "r" (0)); /* ICIALLU: invalidate instruction cache */
+                __asm__("mcr p15, 0, %0, c8, c7, 0\n" :: "r" (0)); /* TLBIALL: invalidate TLB */
+                __asm__("mcr p15, 0, %0, c7, c5, 6\n" :: "r" (0)); /* BPIALL: invalidate branch predictor */
+                __asm__("isb\n");
+
+                /* Enable caches, branch prediction and the MMU */
+                uint32_t sctlr;
+                __asm__("mrc p15, 0, %0, c1, c0, 0" : "=r" (sctlr));
+                sctlr |= SCTLR_ICACHE | SCTLR_BRANCH_PRED | SCTLR_DATA_L2 | SCTLR_MMU;
+                __asm__("mcr p15, 0, %0, c1, c0, 0\n" :: "r" (sctlr));
+                __asm__("isb\n");
+        }
```

Comment:
```suggestion
        /*
         * Check to see if the processor supports VMSAv7 or better, which can
         * be done by testing for if `mmfr0 & 0xf >= 3`.
         */
        __asm__("mrc     p15, #0, r0, c0, c1, #4"); /* mmfr0 */
        __asm__("and     r0, r0, #0xf");
        __asm__("cmp     r0, #3");
        __asm__("blo     .Lend_mmu_setup");
        /*
         * We have to set up an identity map and enable the MMU for caches.
         * Additionally, all page table entries are set to Domain 0, so set up
         * DACR so that Domain zero has permission checks enabled rather than
         * "deny all".
         */

        /* Set DACR Domain 0 permissions checked */
        __asm__("mcr p15, 0, %0, c3, c0, 0\n" :: "r" (1));

        /*
         * Write TTBR
         *
         * No DSB since tables are statically initialized and dcache is off.
         * We or __identity_page_table with 0x3 to set the cacheable flag bits.
         */
        __asm__("mcr p15, 0, %0, c2, c0, 0\n" :: "r" ((uintptr_t)__identity_page_table | 0x3));

        /* Note: we assume Data+L2 cache has been invalidated by reset. */
        __asm__("mcr p15, 0, %0, c7, c5, 0\n" :: "r" (0)); /* ICIALLU: invalidate instruction cache */
        __asm__("mcr p15, 0, %0, c8, c7, 0\n" :: "r" (0)); /* TLBIALL: invalidate TLB */
        __asm__("mcr p15, 0, %0, c7, c5, 6\n" :: "r" (0)); /* BPIALL: invalidate branch predictor */
        __asm__("isb\n");

        /* Enable caches, branch prediction and the MMU */
        __asm__("mrc p15, 0, r0, c1, c0, 0\n"); /* Read SCTLR */
        __asm__("orr r0, r0, %0" ::"r"(SCTLR_ICACHE | SCTLR_BRANCH_PRED |
                                       SCTLR_DATA_L2 | SCTLR_MMU));
        __asm__("mcr p15, 0, r0, c1, c0, 0\n"); /* Write back SCTLR */
        __asm__("isb\n");
        __asm__(".Lend_mmu_setup:\n");

```

The following works in my testing.